File: server/src/services/gapAnalysis.js
// Gap analysis: find periods with browser activity not covered by Harvest entries.

const SESSION_BREAK_MS = 5 * 60 * 1000;  // 5 min gap between heartbeats = new session
const MIN_GAP_MINUTES = 15;               // only flag gaps ≥ 15 minutes

/**
 * Groups raw heartbeats into continuous activity sessions.
 * A session ends when there's a gap > SESSION_BREAK_MS between heartbeats.
 */
export function buildSessions(heartbeats) {
  if (heartbeats.length === 0) return [];

  const sorted = [...heartbeats].sort((a, b) => a.timestamp - b.timestamp);
  const sessions = [];
  let current = {
    start: sorted[0].timestamp,
    end: sorted[0].timestamp,
    domains: new Set([sorted[0].domain]),
    activityTypes: new Set([sorted[0].activity_type || sorted[0].activityType].filter(Boolean))
  };

  for (let i = 1; i < sorted.length; i++) {
    const hb = sorted[i];
    const gap = hb.timestamp - current.end;

    if (gap > SESSION_BREAK_MS) {
      sessions.push(finalizeSession(current));
      current = {
        start: hb.timestamp,
        end: hb.timestamp,
        domains: new Set([hb.domain]),
        activityTypes: new Set([hb.activity_type || hb.activityType].filter(Boolean))
      };
    } else {
      current.end = hb.timestamp;
      if (hb.domain) current.domains.add(hb.domain);
      const type = hb.activity_type || hb.activityType;
      if (type) current.activityTypes.add(type);
    }
  }
  sessions.push(finalizeSession(current));

  return sessions;
}

function finalizeSession(s) {
  return {
    start: s.start,
    end: s.end,
    durationMinutes: Math.round((s.end - s.start) / 60000),
    domains: [...s.domains].filter(Boolean),
    activityTypes: [...s.activityTypes].filter(Boolean)
  };
}

/**
 * Finds gaps: activity sessions not (or only partially) covered by Harvest entries.
 *
 * Harvest entries often only have hours (no start/end time), so we use a
 * time-budget approach: we subtract tracked hours from the day total and
 * flag remaining activity windows as gaps.
 */
export function findGaps(heartbeats, harvestEntries) {
  const sessions = buildSessions(heartbeats);
  if (sessions.length === 0) return [];

  // Build Harvest coverage: if entries have start/end times, use them.
  // Otherwise fall back to flagging all active sessions as potential gaps
  // and subtracting the total tracked hours as an estimate.
  const entriesWithTimes = harvestEntries.filter(e => e.startedTime && e.endedTime);

  let gaps;
  if (entriesWithTimes.length > 0) {
    gaps = findGapsWithTimestamps(sessions, entriesWithTimes);
  } else {
    gaps = findGapsWithBudget(sessions, harvestEntries);
  }

  return gaps.filter(g => (g.end - g.start) / 60000 >= MIN_GAP_MINUTES);
}

function findGapsWithTimestamps(sessions, entries) {
  // Build covered intervals from Harvest start/end times
  const covered = entries.map(e => ({
    start: parseHarvestTime(e.spentDate, e.startedTime),
    end: parseHarvestTime(e.spentDate, e.endedTime)
  })).filter(i => i.start && i.end);

  const gaps = [];
  for (const session of sessions) {
    const uncovered = subtractIntervals(session.start, session.end, covered);
    for (const [start, end] of uncovered) {
      gaps.push({ start, end, domains: session.domains });
    }
  }
  return gaps;
}

function findGapsWithBudget(sessions, harvestEntries) {
  // No timestamps on entries — mark all active sessions as gaps,
  // but trim the most recent ones to account for tracked hours.
  const trackedMs = harvestEntries.reduce((s, e) => s + e.hours * 3600000, 0);
  const totalActiveMs = sessions.reduce((s, sess) => s + (sess.end - sess.start), 0);
  const untrackedMs = Math.max(0, totalActiveMs - trackedMs);

  if (untrackedMs < MIN_GAP_MINUTES * 60000) return [];

  // Return all sessions as gaps (user has to decide which ones to log)
  return sessions.map(sess => ({
    start: sess.start,
    end: sess.end,
    domains: sess.domains
  }));
}

/**
 * Subtracts covered intervals [start,end][] from the range [from, to].
 * Returns uncovered sub-ranges as [start, end][] pairs.
 */
function subtractIntervals(from, to, covered) {
  const relevant = covered
    .filter(i => i.end > from && i.start < to)
    .sort((a, b) => a.start - b.start);

  const result = [];
  let cursor = from;

  for (const interval of relevant) {
    if (interval.start > cursor) {
      result.push([cursor, Math.min(interval.start, to)]);
    }
    cursor = Math.max(cursor, interval.end);
    if (cursor >= to) break;
  }

  if (cursor < to) result.push([cursor, to]);
  return result;
}

function parseHarvestTime(date, timeStr) {
  if (!date || !timeStr) return null;
  try {
    return new Date(`${date}T${timeStr}`).getTime();
  } catch {
    return null;
  }
}

File: server/gap_analysis.py
from datetime import datetime
from typing import Optional

SESSION_BREAK_MS = 5 * 60 * 1000   # 5 minuten = nieuw blok
MIN_GAP_MINUTES  = 15               # gaten korter dan 15 min negeren

def build_sessions(heartbeats: list) -> list:
    if not heartbeats:
        return []

    sorted_hb = sorted(heartbeats, key=lambda h: h["timestamp"])

    sessions = []
    current = {
        "start":         sorted_hb[0]["timestamp"],
        "end":           sorted_hb[0]["timestamp"],
        "domains":       {sorted_hb[0]["domain"]} if sorted_hb[0].get("domain") else set(),
        "activityTypes": {sorted_hb[0]["activity_type"]} if sorted_hb[0].get("activity_type") else set(),
    }

    for hb in sorted_hb[1:]:
        gap = hb["timestamp"] - current["end"]
        if gap > SESSION_BREAK_MS:
            sessions.append(_finalize(current))
            current = {
                "start":         hb["timestamp"],
                "end":           hb["timestamp"],
                "domains":       {hb["domain"]} if hb.get("domain") else set(),
                "activityTypes": {hb["activity_type"]} if hb.get("activity_type") else set(),
            }
        else:
            current["end"] = hb["timestamp"]
            if hb.get("domain"):        current["domains"].add(hb["domain"])
            if hb.get("activity_type"): current["activityTypes"].add(hb["activity_type"])

    sessions.append(_finalize(current))
    return sessions

def _finalize(s: dict) -> dict:
    return {
        "start":           s["start"],
        "end":             s["end"],
        "durationMinutes": round((s["end"] - s["start"]) / 60000),
        "domains":         [d for d in s["domains"] if d],
        "activityTypes":   [t for t in s["activityTypes"] if t],
    }

def find_gaps(heartbeats: list, harvest_entries: list) -> list:
    sessions = build_sessions(heartbeats)
    if not sessions:
        return []

    entries_with_times = [
        e for e in harvest_entries
        if e.get("startedTime") and e.get("endedTime")
    ]

    if entries_with_times:
        gaps = _gaps_with_timestamps(sessions, entries_with_times)
    else:
        gaps = _gaps_with_budget(sessions, harvest_entries)

    return [g for g in gaps if (g["end"] - g["start"]) / 60000 >= MIN_GAP_MINUTES]

def _gaps_with_timestamps(sessions: list, entries: list) -> list:
    covered = []
    for e in entries:
        start = _parse_harvest_time(e["spentDate"], e["startedTime"])
        end   = _parse_harvest_time(e["spentDate"], e["endedTime"])
        if start and end:
            covered.append((start, end))

    gaps = []
    for sess in sessions:
        for (start, end) in _subtract_intervals(sess["start"], sess["end"], covered):
            gaps.append({"start": start, "end": end, "domains": sess["domains"]})
    return gaps

def _gaps_with_budget(sessions: list, entries: list) -> list:
    tracked_ms   = sum(e["hours"] * 3_600_000 for e in entries)
    total_active = sum(s["end"] - s["start"] for s in sessions)
    untracked    = max(0, total_active - tracked_ms)

    if untracked < MIN_GAP_MINUTES * 60_000:
        return []

    return [{"start": s["start"], "end": s["end"], "domains": s["domains"]} for s in sessions]

def _subtract_intervals(from_ts: int, to_ts: int, covered: list) -> list:
    relevant = sorted(
        [(s, e) for (s, e) in covered if e > from_ts and s < to_ts],
        key=lambda x: x[0],
    )
    result, cursor = [], from_ts
    for (s, e) in relevant:
        if s > cursor:
            result.append((cursor, min(s, to_ts)))
        cursor = max(cursor, e)
        if cursor >= to_ts:
            break
    if cursor < to_ts:
        result.append((cursor, to_ts))
    return result

def _parse_harvest_time(date_str: Optional[str], time_str: Optional[str]) -> Optional[int]:
    if not date_str or not time_str:
        return None
    try:
        dt = datetime.fromisoformat(f"{date_str}T{time_str}")
        return int(dt.timestamp() * 1000)
    except ValueError:
        return None
